File: sdk-java/src/main/java/com/crosscert/passkey/sdk/PasskeyClientConfig.java
package com.crosscert.passkey.sdk;

import org.slf4j.MDC;
import java.net.URI;
import java.time.Clock;
import java.time.Duration;
import java.util.Objects;
import java.util.function.Supplier;

/**
 * SDK 클라이언트 설정. 필수값(baseUrl, apiKeySupplier)은 {@link #builder} 인자로
 * 강제되고, 선택값은 미설정 시 기본값으로 치환된다.
 *
 * apiKeySupplier/traceIdProvider 의 반환값 null/blank 검증은 생성 시점이 아니라
 * 매 요청 시점(인터셉터)에서 한다 — 키가 도중에 비워질 수 있으므로.
 */
public final class PasskeyClientConfig {

    public static final String MDC_TRACE_ID_KEY = "traceId";

    private final URI baseUrl;
    private final Supplier<String> apiKeySupplier;
    private final Duration connectTimeout;
    private final Duration readTimeout;
    private final Duration jwksCacheTtl;
    private final Clock clock;
    private final Supplier<String> traceIdProvider;

    private PasskeyClientConfig(Builder b) {
        this.baseUrl = Objects.requireNonNull(b.baseUrl, "baseUrl");
        this.apiKeySupplier = Objects.requireNonNull(b.apiKeySupplier, "apiKeySupplier");
        this.connectTimeout = (b.connectTimeout != null) ? b.connectTimeout : Duration.ofSeconds(3);
        this.readTimeout = (b.readTimeout != null) ? b.readTimeout : Duration.ofSeconds(10);
        this.jwksCacheTtl = (b.jwksCacheTtl != null) ? b.jwksCacheTtl : Duration.ofMinutes(5);
        this.clock = (b.clock != null) ? b.clock : Clock.systemUTC();
        this.traceIdProvider = (b.traceIdProvider != null)
                ? b.traceIdProvider
                : () -> MDC.get(MDC_TRACE_ID_KEY);
    }

    public URI baseUrl() { return baseUrl; }
    public Supplier<String> apiKeySupplier() { return apiKeySupplier; }
    public Duration connectTimeout() { return connectTimeout; }
    public Duration readTimeout() { return readTimeout; }
    public Duration jwksCacheTtl() { return jwksCacheTtl; }
    public Clock clock() { return clock; }
    public Supplier<String> traceIdProvider() { return traceIdProvider; }

    public static Builder builder(URI baseUrl, Supplier<String> apiKeySupplier) {
        return new Builder(baseUrl, apiKeySupplier);
    }

    /** 모든 선택값에 기본값을 적용한 설정. (기존 외부 소비자/테스트 호환) */
    public static PasskeyClientConfig defaults(URI baseUrl, Supplier<String> apiKeySupplier) {
        return builder(baseUrl, apiKeySupplier).build();
    }

    public static final class Builder {
        private final URI baseUrl;
        private final Supplier<String> apiKeySupplier;
        private Duration connectTimeout;
        private Duration readTimeout;
        private Duration jwksCacheTtl;
        private Clock clock;
        private Supplier<String> traceIdProvider;

        private Builder(URI baseUrl, Supplier<String> apiKeySupplier) {
            this.baseUrl = baseUrl;
            this.apiKeySupplier = apiKeySupplier;
        }

        /** null 이면 기본값(3s) 적용. */
        public Builder connectTimeout(Duration v) { this.connectTimeout = v; return this; }
        /** null 이면 기본값(10s) 적용. */
        public Builder readTimeout(Duration v) { this.readTimeout = v; return this; }
        /** null 이면 기본값(5m) 적용. */
        public Builder jwksCacheTtl(Duration v) { this.jwksCacheTtl = v; return this; }
        /** null 이면 기본값(systemUTC) 적용. */
        public Builder clock(Clock v) { this.clock = v; return this; }
        /** null 이면 기본값(MDC.get("traceId")) 적용. */
        public Builder traceIdProvider(Supplier<String> v) { this.traceIdProvider = v; return this; }

        public PasskeyClientConfig build() { return new PasskeyClientConfig(this); }
    }
}

File: sdk-java/src/main/kotlin/com/crosscert/passkey/sdk/PasskeyClientConfig.kt
package com.crosscert.passkey.sdk

import org.slf4j.MDC
import java.net.URI
import java.time.Clock
import java.time.Duration
import java.util.Objects
import java.util.function.Supplier

/**
 * SDK 클라이언트 설정.
 *
 * 외부 Java 소비자 호환: 원본 Java `record` 의 정규 생성자가 선택값에 `null` 을
 * 받아 기본값으로 치환하던 동작을 그대로 보존한다. 즉
 * `new PasskeyClientConfig(base, supplier, null, null, null, null, null)` 처럼
 * 일부/전부 선택값을 null 로 넘기면 각각 기본값이 적용된다(원본 컴팩트 생성자와 동일).
 * 접근자는 record 식(`config.baseUrl()` 등) 으로 유지된다([JvmName]).
 *
 * apiKeySupplier 의 반환 키 null/blank 검증은 생성 시점이 아니라 매 요청 시점
 * (RedactingRequestInterceptor)에서 한다 — 키가 도중에 비워질 수 있으므로.
 */
class PasskeyClientConfig
@JvmOverloads
constructor(
    baseUrl: URI,
    apiKeySupplier: Supplier<String?>,
    connectTimeout: Duration? = null,
    readTimeout: Duration? = null,
    jwksCacheTtl: Duration? = null,
    clock: Clock? = null,
    traceIdProvider: Supplier<String?>? = null,
) {
    @get:JvmName("baseUrl")
    val baseUrl: URI = Objects.requireNonNull(baseUrl)

    @get:JvmName("apiKeySupplier")
    val apiKeySupplier: Supplier<String?> = Objects.requireNonNull(apiKeySupplier)

    @get:JvmName("connectTimeout")
    val connectTimeout: Duration = connectTimeout ?: Duration.ofSeconds(3)

    @get:JvmName("readTimeout")
    val readTimeout: Duration = readTimeout ?: Duration.ofSeconds(10)

    @get:JvmName("jwksCacheTtl")
    val jwksCacheTtl: Duration = jwksCacheTtl ?: Duration.ofMinutes(5)

    @get:JvmName("clock")
    val clock: Clock = clock ?: Clock.systemUTC()

    @get:JvmName("traceIdProvider")
    val traceIdProvider: Supplier<String?> = traceIdProvider ?: Supplier { MDC.get(MDC_TRACE_ID_KEY) }

    // 원본 Java record 의 값 동등성/해시/toString 을 보존(외부 소비자가 비교/로깅할 수 있음).
    // Supplier/Clock 은 의미 있는 value equality 가 없어 record 와 동일하게 인스턴스 동일성으로 비교된다.
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PasskeyClientConfig) return false
        return baseUrl == other.baseUrl &&
            apiKeySupplier == other.apiKeySupplier &&
            connectTimeout == other.connectTimeout &&
            readTimeout == other.readTimeout &&
            jwksCacheTtl == other.jwksCacheTtl &&
            clock == other.clock &&
            traceIdProvider == other.traceIdProvider
    }

    override fun hashCode(): Int =
        Objects.hash(baseUrl, apiKeySupplier, connectTimeout, readTimeout, jwksCacheTtl, clock, traceIdProvider)

    override fun toString(): String =
        "PasskeyClientConfig[baseUrl=$baseUrl, apiKeySupplier=$apiKeySupplier, " +
            "connectTimeout=$connectTimeout, readTimeout=$readTimeout, jwksCacheTtl=$jwksCacheTtl, " +
            "clock=$clock, traceIdProvider=$traceIdProvider]"

    companion object {
        const val MDC_TRACE_ID_KEY: String = "traceId"

        @JvmStatic
        fun defaults(baseUrl: URI, apiKeySupplier: Supplier<String?>): PasskeyClientConfig =
            PasskeyClientConfig(baseUrl, apiKeySupplier)
    }
}
